### ISSUE TITLE
max操作反向传播没有梯度

### ISSUE THREAD

**Issue by qizhen816** [shuaizzZ/Recall-Loss-PyTorch#1]
`_, predict = torch.max(input, 1)# # (N, C, *) ==> (N, 1, *)`

这里得到的下标predict是没有梯度的，训练会报错，请问你是怎么解决的呢

**Comment by bithuanglq** [comment]
附议

**Comment by bithuanglq** [comment]
> `_, predict = torch.max(input, 1)# # (N, C, *) ==> (N, 1, *)`
> 
> 这里得到的下标predict是没有梯度的，训练会报错，请问你是怎么解决的呢

我也是

**Comment by shuaizzZ** [comment]
> `_, predict = torch.max(input, 1)# # (N, C, *) ==> (N, 1, *)`
> 
> 这里得到的下标predict是没有梯度的，训练会报错，请问你是怎么解决的呢

可以试下这个：
    def forward(self, input, target):
        N, C = input.size()[:2]
        logpt = F.log_softmax(input, dim=1)
        pt = logpt.exp()

        ## convert target (N, 1, *) into one hot vector (N, C, *)
        target = target.view(N, 1, -1)  # (N, 1, *)
        target_onehot = torch.zeros(pt.size()).type_as(pt) # (N, 1, *) ==> (N, C, *)
        target_onehot.scatter_(1, target, 1) # (N, C, *)

        true_positive = torch.sum(pt * target_onehot, dim=2)  # (N, C)
        total_target = torch.sum(target_onehot, dim=2)  # (N, C)
        ## Recall = TP / (TP + FN)
        recall = (true_positive + self.smooth) / (total_target + self.smooth)  # (N, C)

        if hasattr(self, 'weight'):
            if self.weight.type() != input.type():
                self.weight = self.weight.type_as(input)
                recall = recall * self.weight * C  # (N, C)
        recall_loss = 1 - torch.mean(recall)  # 1

        return recall_loss

**Comment by shuaizzZ** [comment]
def forward(self, input, target):
        N, C = input.size()[:2]
        logpt = F.log_softmax(input, dim=1)
        pt = logpt.exp()

        ## convert target (N, 1, *) into one hot vector (N, C, *)
        target = target.view(N, 1, -1)  # (N, 1, *)
        target_onehot = torch.zeros(pt.size()).type_as(pt) # (N, 1, *) ==> (N, C, *)
        target_onehot.scatter_(1, target, 1) # (N, C, *)

        true_positive = torch.sum(pt * target_onehot, dim=2)  # (N, C)
        total_target = torch.sum(target_onehot, dim=2)  # (N, C)
        ## Recall = TP / (TP + FN)
        recall = (true_positive + self.smooth) / (total_target + self.smooth)  # (N, C)

        if hasattr(self, 'weight'):
            if self.weight.type() != input.type():
                self.weight = self.weight.type_as(input)
                recall = recall * self.weight * C  # (N, C)
        recall_loss = 1 - torch.mean(recall)  # 1

        return recall_loss

**Comment by shuaizzZ** [comment]
可以试下上面的代码，不过我自己很早之前试过，这个loss似乎没啥用

**Comment by bithuanglq** [comment]
好的谢谢！确实效果没有论文里那么好



---原始邮件---
发件人: ***@***.***&gt;
发送时间: 2022年4月2日(周六) 下午5:16
收件人: ***@***.***&gt;;
抄送: ***@***.******@***.***&gt;;
主题: Re: [shuaizzZ/Recall-Loss-PyTorch] max操作反向传播没有梯度 (#1)




 
可以试下上面的代码，不过我自己很早之前试过，这个loss似乎没啥用
